### PR TITLE
[WIP] Starting to attempt to improve documentation.

### PR DESCRIPTION
**Pull request by icefoxen**
Mostly just some basic cleanup for the *few* bits I actually understand.  This is presented mainly as a place for me to put ongoing work and get comments/discussion.  I realize now that I actually understand very little of gfx-hal, so it's going to be slow going, but I'll do what I can as I figure it out.

My general plan:
 * Start with just filling out API docs where possible
 * Write up a guide explaining HAL structure and how to make the pieces fit together, like my LearningGfx tutorial does for pre-ll
 * Maybe essentially dissect the `gfx_ocean` program in detail as a case study, if I can figure out how the heck it works and I still have energy.  :-)

This isn't going to be super fast, but I'll do what I can!

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds (might be temporarily impossible, WIP)
- [x] tested examples with the following backends: n/a

### CODE REVIEW

File: src/hal/src/image.rs [hunk 1]
```diff
@@ -246,10 +246,11 @@ pub const CUBE_FACES: [CubeFace; 6] = [
 #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum Kind {
-    /// A single row of texels.
+    /// A single one-dimensional row of texels.
```
Comment (omni-viral) [inline src/hal/src/image.rs]: Tbh, I've never seen two-dimentional rows.

File: src/hal/src/format.rs [hunk 2]
```diff
@@ -2,7 +2,11 @@
 //! Applicable to textures, views, and vertex buffers.
 
 bitflags!(
-    ///
+    /// Bitflags that describes what properties of a buffer
```
Comment (kvark) [inline src/hal/src/format.rs]: "describe"

Comment (kvark) [inline src/hal/src/format.rs]: "buffer" -> "image"

File: src/hal/src/image.rs [hunk 3]
```diff
@@ -368,8 +370,10 @@ impl Usage {
 #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum WrapMode {
-    /// Tile the texture. That is, sample the coordinate modulo `1.0`. This is
-    /// the default.
+    /// Tile the texture, that is, sample the coordinate modulo `1.0`, so 
+    /// addressing the texture beyond an edge will "wrap" back from the
+    /// other edge.
+    /// This is the default.
```
Comment (kvark) [inline src/hal/src/image.rs]: I don't think there is a default any more

File: src/hal/src/image.rs [hunk 4]
```diff
@@ -396,7 +400,7 @@ impl Into<f32> for Lod {
     }
 }
 
-/// A wrapper for the 8bpp RGBA color, encoded as u32.
+/// A wrapper for an RGBA color with 8 bits per pixel, encoded as a u32.
```
Comment (kvark) [inline src/hal/src/image.rs]: in the long run, we should tighten up the use of "pixel" versus "texel" versus "fragment".

File: src/hal/src/lib.rs [hunk 5]
```diff
@@ -304,7 +305,9 @@ pub trait Instance {
     fn enumerate_adapters(&self) -> Vec<Adapter<Self::Backend>>;
 }
 
-/// Different types of a specific API.
+/// The `Backend` trait wraps together all the types needed
+/// to implement a graphics backend.  Each backend, such as OpenGL
```
Comment (kvark) [inline src/hal/src/lib.rs]: the second sentence seems very repetit-ory. Perhaps, something like this would sound better?
> The `Backend` trait wraps together all the types needed to implement a graphics backend, turning into concrete types on OpenGL, Metal, etc.

File: src/hal/src/lib.rs [hunk 6]
```diff
@@ -366,12 +369,13 @@ impl Error for SubmissionError {
 pub type SubmissionResult<T> = Result<T, SubmissionError>;
 
 
-/// Represents a handle to a physical device.
+/// Represents a handle to a physical device, such as a particular
```
Comment (kvark) [inline src/hal/src/lib.rs]: This needs to be fixed. We have a separate type `PhysicalDevice` now, and the `Gpu` struct is not that but rather a combination of the logical device and the hardware queues it provides.

File: src/hal/src/window.rs [hunk 7]
```diff
@@ -154,7 +154,20 @@ pub enum FrameSync<'a, B: Backend> {
     Fence(&'a B::Fence),
 }
 
-/// Allows you to configure a `Swapchain` for creation.
+/// Contains all the data necessary to create a new `SwapChain`:
```
Comment (kvark) [inline src/hal/src/window.rs]: "SwapChain" -> "Swapchain"

File: src/hal/src/image.rs (hunk 4 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: Which is it in this case?  Texel?

File: src/hal/src/format.rs (hunk 2 above)
Comment (icefoxen) [reply src/hal/src/format.rs]: But these properties are applicable to vertex buffers as well, as per the module docs?

File: src/hal/src/image.rs (hunk 1 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: Yeah but the language used for that section in general is a little squirrelly and I'm trying to make it consistent (maybe unsuccessfully).  We have three "axes" this type can vary on: Dimension (1,2,3), array/not array, and cube/not cube, with dimension only valid for not-cube.

File: src/hal/src/image.rs (hunk 3 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: Is there a concrete default then?

File: src/hal/src/format.rs (hunk 2 above)
Comment (kvark) [reply src/hal/src/format.rs]: Vertex buffers can't be COLOR/DEPTH/STENCIL, which is what `AspectFlags` is about

File: src/hal/src/lib.rs (hunk 5 above)
Comment (icefoxen) [reply src/hal/src/lib.rs]: Yeah, it could be phrased better.  This is one of the Important Bits though, so I feel like the repetition is useful to make sure the relationship between the trait, the backend modules, and the backend API's are clear.  I'll rework it.

File: src/hal/src/lib.rs (hunk 6 above)
Comment (icefoxen) [reply src/hal/src/lib.rs]: Ah, interesting, I need to look at this more.  Thank you.

File: src/hal/src/image.rs (hunk 4 above)
Comment (kvark) [reply src/hal/src/image.rs]: AFAIK, this is the distinction:
  - "texel" refers to the actual contents of an image, which is the case here
  - "pixel" refers to the *result* on screen
  - "fragment" refers to the unit of work for the shader, multiple fragments can contribute to the same pixels

File: src/hal/src/image.rs (hunk 3 above)
Comment (kvark) [reply src/hal/src/image.rs]: Not really. Neither `WrapMode` or `SamplerInfo` (which the `WrapMode` is a part of) implement `Default`, and the only constructor of the former gets `WrapMode` as an argument.

File: src/hal/src/command/mod.rs [hunk 8]
```diff
@@ -19,22 +19,33 @@ use std::borrow::Borrow;
...
+///
+/// Vulkan describes a primary command buffer as one which can be directly submitted
+/// to a queue, and can execute `Secondary` command buffers.
 pub enum Primary { }
 impl Level for Primary { }
+
 /// Indicates a secondary command buffer.
+///
+/// Vulkan describes a secondary command buffer as one which cannot be directly submitted
+/// to a queue, but can be executed by a primary command buffer.  This allows you to
```
Comment (kvark) [inline src/hal/src/command/mod.rs]: "This allows you" - I think needs to be rephrased to avoid "you". Also, there is an extra space before it.

File: src/hal/src/command/mod.rs [hunk 9]
```diff
@@ -64,7 +75,7 @@ unsafe impl<'a, B: Backend, C, L: Level> Submittable<'a, B, C, L> for &'a Submit
     unsafe fn as_buffer(self) -> &'a B::CommandBuffer { &self.0 }
 }
 
-/// A convenience for not typing out the full signature of a secondary command buffer.
+/// A convenience type for not typing out the full signature of a secondary command buffer.
```
Comment (kvark) [inline src/hal/src/command/mod.rs]: maybe `A convenience alias for not typing ...` so that we don't repeat `type`

File: src/hal/src/command/transfer.rs [hunk 10]
```diff
@@ -9,63 +9,68 @@ use queue::capability::{Supports, Transfer};
...
     pub y: i32,
-    ///
+    /// Z offset, in texels.
     pub z: i32,
 }
 
-/// Region of two buffers for copying.
+/// Specifies a source region and a destination
+/// region in a buffer for copying.
+/// DOC TODO: What units for src, dst, size?
```
Comment (kvark) [inline src/hal/src/command/transfer.rs]: bytes for life!

File: src/hal/src/command/transfer.rs [hunk 11]
```diff
@@ -9,63 +9,68 @@ use queue::capability::{Supports, Transfer};
...
+/// Bundles together all the data need to copy data from one `Image`
+/// to another.
 #[derive(Clone, Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct ImageCopy {
-    ///
+    /// The aspect mask of what to copy: color, depth and/or stencil information.
     pub aspect_mask: format::AspectFlags,
-    ///
+    /// The image subresource to copy from
```
Comment (kvark) [inline src/hal/src/command/transfer.rs]: let's end comments with a dot consistently

File: src/hal/src/lib.rs [hunk 12]
```diff
@@ -306,7 +306,7 @@ pub trait Instance {
 }
 
 /// The `Backend` trait wraps together all the types needed
-/// to implement a graphics backend.  Each backend, such as OpenGL
+/// for a graphics backend.  Each backend module, such as OpenGL
```
Comment (kvark) [inline src/hal/src/lib.rs]: double space found

File: src/hal/src/pso/graphics.rs [hunk 13]
```diff
@@ -6,44 +6,55 @@ use super::input_assembler::{AttributeDesc, InputAssemblerDesc, VertexBufferDesc
 use super::output_merger::{ColorBlendDesc, DepthStencilDesc};
 
 /// A complete set of shaders to build a graphics pipeline.
+///
+/// All except the vertex shader are optional.
+/// DOC TODO: What happens if, say, a fragment shader is not defined?
```
Comment (kvark) [inline src/hal/src/pso/graphics.rs]: > If a fragment shader is omitted, the results of fragment processing are undefined. Specifically, any fragment color outputs are considered to have undefined values, and the fragment depth is considered to be unmodified. This can be useful for depth-only rendering.

File: src/hal/src/pso/graphics.rs [hunk 14]
```diff
@@ -6,44 +6,55 @@ use super::input_assembler::{AttributeDesc, InputAssemblerDesc, VertexBufferDesc
...
+    /// be parameters for creating geometry) and produces one or more
+    /// output patches.
     pub hull: Option<EntryPoint<'a, B>>,
-    ///
+    /// A shader that takes in domains produced from a hull shader's output
+    /// patches and computes actual vertex positions.
     pub domain: Option<EntryPoint<'a, B>>,
-    ///
+    /// A shader that takes given input vertexes and outputs one
+    /// or more output vertexes.
```
Comment (kvark) [inline src/hal/src/pso/graphics.rs]: can technically output 0 vertices as well, which is used for GPU-based culling

File: src/hal/src/pso/graphics.rs [hunk 15]
```diff
@@ -6,44 +6,55 @@ use super::input_assembler::{AttributeDesc, InputAssemblerDesc, VertexBufferDesc
...
     pub geometry: Option<EntryPoint<'a, B>>,
-    ///
+    /// A shader that outputs a value for a texel.
```
Comment (kvark) [inline src/hal/src/pso/graphics.rs]: for a fragment!

File: src/hal/src/pso/graphics.rs [hunk 16]
```diff
@@ -104,15 +121,16 @@ pub enum FrontFace {
     CounterClockwise,
 }
 
-///
+/// A depth bias allows you to specify a hint for how to draw
```
Comment (kvark) [inline src/hal/src/pso/graphics.rs]: I'd say "A depth bias allows changing the produced depth values for fragments slightly but consistently, allowing to draw multiple layers of the same geometry without Z-fighting"

File: src/hal/src/pso/graphics.rs [hunk 17]
```diff
@@ -147,23 +165,23 @@ impl Rasterizer {
     };
 }
 
-///
+/// A description of an equation for how to blend transparent, overlapping texels.
```
Comment (kvark) [inline src/hal/src/pso/graphics.rs]: fragments :)

File: src/hal/src/pso/mod.rs [hunk 18]
```diff
@@ -206,6 +206,7 @@ pub struct Specialization {
 }
 
 /// Scalar specialization constant with value for overriding.
+/// DOC TODO, don't know what this means.
```
Comment (kvark) [inline src/hal/src/pso/mod.rs]: specialization constants allow for easy configuration of multiple similar pipelines, e.g. you may have a boolean exposed to the shader that switches the specularity on/off, and provide it as a specialization constant. That would make on/off to be separate PSOs but they share most of the internal stuff and are fast to produce. More importantly, they are fast to execute, since the driver can optimize out the branch on that other PSO creation.

File: src/hal/src/pso/output_merger.rs [hunk 19]
```diff
@@ -98,36 +98,39 @@ pub enum BlendOp {
 }
 
 impl BlendOp {
-    ///
+    /// Replaces the destination value with the source.
```
Comment (kvark) [inline src/hal/src/pso/output_merger.rs]: "Replaces" -> "Replace" for consistency

File: src/hal/src/queue/submission.rs [hunk 20]
```diff
@@ -24,7 +24,8 @@ where
     pub signal_semaphores: &'a [&'a B::Semaphore],
 }
 
-/// Submission information for a command queue.
+/// Submission information for a command queue, generic on a particular
```
Comment (kvark) [inline src/hal/src/queue/submission.rs]: "generic on" -> "generic over"

File: src/hal/src/queue/submission.rs [hunk 21]
```diff
@@ -101,7 +102,8 @@ where
         }
     }
 
-    /// Promote a submission to a higher capability type.
+    /// Promote a submission to a higher capability type, so you can use this to
```
Comment (kvark) [inline src/hal/src/queue/submission.rs]: similarly, we should probably rephrase this to avoid "you"

File: src/hal/src/lib.rs (hunk 12 above)
Comment (icefoxen) [reply src/hal/src/lib.rs]: Unfortunately I was taught that you always put two spaces between sentences.  Markdown and HTML don't care but it's a hard habit to break.  I'll try to keep it consistent with your style.

File: src/hal/src/adapter.rs [hunk 22]
```diff
@@ -121,7 +126,7 @@ impl<B: Backend> Adapter<B> {
     /// # fn main() {
     ///
     /// # let adapter: hal::Adapter<empty::Backend> = return;
-    /// let gpu = adapter.open_with::<_, General>(1, |_| true);
+    /// let device = adapter.open_with::<_, General>(1, |_| true);
```
Comment (kvark) [inline src/hal/src/adapter.rs]: `let (device, queues) = adapter.open_with::<_, General>(1, |_| true).unwrap();`

File: src/hal/src/command/mod.rs [hunk 23]
```diff
@@ -43,9 +43,9 @@ impl Level for Primary { }
 /// Indicates a secondary command buffer.
 ///
 /// Vulkan describes a secondary command buffer as one which cannot be directly submitted
-/// to a queue, but can be executed by a primary command buffer.  This allows you to
-/// construct multiple secondary command buffers that do specific things, and compose
-/// them together into primary command buffers.
+/// to a queue, but can be executed by a primary command buffer.  This allows
+/// multiple secondary command buffers to be constructed which do specific
```
Comment (kvark) [inline src/hal/src/command/mod.rs]: missing comma before `which`?

File: src/hal/src/device.rs [hunk 24]
```diff
@@ -1,6 +1,15 @@
...
+//! This module exposes the `Device` trait, which provides methods for creating
+//! and managing 
+//!
+//! The `Adapter` and `Device` are very similar to the Vulkan concept of 
+//! "physical devices" vs. "logical devices"; an `Adapter` is single GPU 
+//! (or CPU) that implements a backend, a `Device` is a 
+//! handle to that GPU that fulfills the capabilities you request.
+
+
+//! used for creating and managing graphics resources, and
```
Comment (kvark) [inline src/hal/src/device.rs]: "Used"

File: src/hal/src/pso/mod.rs [hunk 25]
```diff
@@ -206,21 +215,20 @@ pub struct Specialization {
 }
 
 /// Scalar specialization constant with value for overriding.
-/// DOC TODO, don't know what this means.
 #[derive(Debug, Clone)]
 pub enum Constant {
-    ///
+    /// `bool` value.
```
Comment (kvark) [inline src/hal/src/pso/mod.rs]: heh, might want to just add `missing_docs` here, it's fairly self-descriptive

File: src/hal/src/queue/mod.rs [hunk 26]
```diff
@@ -1,12 +1,11 @@
-/*! Command queues.
-
-    Queues are the execution paths of the graphical processing units. These process
-    submitted commands buffers.
-
-    There are different types of queues, which can only handle associated command buffers.
-    `CommandQueue<B, C>` has the capability defined by `C` for backend `B`: graphics, 
-    compute and transfer.
-!*/
+//! Command queues.
```
Comment (kvark) [inline src/hal/src/queue/mod.rs]: is that a new standard - to use single-line comments for the module docs?

Comment (icefoxen) [reply src/hal/src/queue/mod.rs]: It's just what I've seen everywhere else...

File: src/hal/src/pso/mod.rs (hunk 25 above)
Comment (icefoxen) [reply src/hal/src/pso/mod.rs]: 👍

**Comment by kvark** [comment]
@icefoxen I assume this needs to go in fast in order to avoid collision with #1837 . Please rebase?

**Comment by icefoxen** [comment]
Rebased but I'm really not sure it improved anything over a merge.  :roll_eyes: Sorry, I really do suck with git.

**Comment by icefoxen** [comment]
Closing in favor of #1862
